sycl/doc/extensions/KernelProperties/KernelProperties.asciidoc

```diff
@@ -0,0 +1,519 @@
+
+Jessica Davies, Intel +
+Joe Garvey, Intel +
+Greg Lueck, Intel +
+John Pennycook, Intel +
+Roland Schulz, Intel
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, Revision 3 and
+the following extensions:
+
+- SYCL_EXT_ONEAPI_PROPERTY_LIST
+
+== Feature Test Macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros".  Therefore, an
+implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_KERNEL_PROPERTIES` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's APIs the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version.  Base features are supported.
+|===
+
+== Overview
+
+SYCL 2020 allows for attributes to be applied to the function-type of kernel
+function declarations. These attributes are often used to pass information
+to a backend device compiler.
+
+There are several known disadvantages to relying on attributes for such
+information, including:
+
+- Host {cpp} compilers are free to ignore unrecognized attributes, implying
+  that attributes should not be employed to convey information that cannot be
+  ignored. Many of the attributes in SYCL 2020 convey information that cannot
+  be ignored (e.g. a kernel may only execute correctly with a specific
+  sub-group size, or on devices which have specific aspects).
+
+- Library-only implementations of SYCL cannot reason about attributes at all.
+
+- SMCP implementations of SYCL must ensure that the host and device compiler
+  both understand and agree upon the meaning of each attribute.
+
+- It is complicated (although not impossible) to generate multiple variants of
+  a kernel with different combinations of attributes.
+
+- There is no way to inspect the properties associated with a kernel at
+  compile-time (e.g. via type traits or similar); any such inspection must be
+  performed at run-time and *after* compiling the kernel.
+
+This extension proposes a replacement for these kernel attributes, in the form
+of a `property_list` accepting properties with compile-time constant
+values, to address several of these issues.
+
+== Kernel Properties
+
+The kernel properties below correspond to kernel attributes defined in
+Section 5.8.1 of the SYCL 2020 specification.  Note that deprecated attributes
+(such as `vec_type_hint`) are not included.  The `requires` attribute has been
+renamed `device_has`, to avoid future conflicts with the {cpp}20 keyword.
+
+```c++
+namespace sycl {
+namespace ext {
+namespace oneapi {
+
+// Corresponds to reqd_work_group_size
+struct work_group_size {
+  template <size_t... Dims>
+  using value_t = property_value<work_group_size, std::integral_constant<size_t, Dims>...>;
```

Comment:
```suggestion
private:
template <size_t... Dims>
struct work_group_size_t {
  /// Unspecified implementation detail
```
and actually we do not even need to expose this type at all in the specification.
Probably a CRTP-based implementation would be used.